components/nodemanager-service/pgdb/project_update.go

```diff
@@ -0,0 +1,267 @@
+package pgdb
+
+import (
+	"context"
+
+	"encoding/json"
+
+	iam_v2 "github.com/chef/automate/api/interservice/authz/v2"
+	"github.com/chef/automate/components/nodemanager-service/api/nodes"
+	project_update_lib "github.com/chef/automate/lib/authz"
+	"github.com/chef/automate/lib/stringutils"
+	uuid "github.com/chef/automate/lib/uuid4"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+const selectNodesProjectData = `
+SELECT
+  n.id,
+  n.projects_data
+FROM nodes n
+`
+
+type NodeProjectData struct {
+	ID                   string `db:"id"`
+	Data                 string `db:"projects_data"`
+	ProjectDataKeyValues []*nodes.ProjectsData
+}
+
+// JobCancel - cancel the a currently running project update
+func (db *DB) JobCancel(ctx context.Context, projectID string) error {
+	log.Debugf("Node manager project update JobCancel %s", projectID)
+	if db.ProjectUpdate.ID == projectID {
+		db.ProjectUpdate.running = false
+	}
+	return nil
+}
+
+// UpdateProjectTags - start a project update
+func (db *DB) UpdateProjectTags(ctx context.Context,
+	projectRules map[string]*iam_v2.ProjectRules) ([]string, error) {
+	jobUUID, _ := uuid.NewV4()
+	log.Debugf("Running node manager UpdateProjectTags %v ID: %s", projectRules, jobUUID.String())
+
+	// setup for new project update run
+	db.ProjectUpdate.jobStatusError = nil
+	db.ProjectUpdate.running = true
+	db.ProjectUpdate.jobStatus = project_update_lib.JobStatus{
+		Completed:             false,
+		PercentageComplete:    0,
+		EstimatedEndTimeInSec: 0,
+	}
+	db.ProjectUpdate.ID = jobUUID.String()
```

Comment:
This code knows a lot about the ProjectUpdate struct, often specifying the same values for the same fields, etc. It might make sense to move a lot of the ProjectUpdate state manipulation where that struct is defined, for example, here this is resetting the ProjectUpdate to some initial state which could be encapsulated with the struct definition.

Reply:
In particular, code like this block is repeated several times and could be a constant where the struct is defined.

```
project_update_lib.JobStatus{
		Completed:             false,
		PercentageComplete:    0,
		EstimatedEndTimeInSec: 0,
	}
```